src/os_net/os_net.c

```diff
@@ -40,64 +40,58 @@ static socklen_t us_l = sizeof(n_us);
 
 
 /* Bind a specific port */
-static int OS_Bindport(u_int16_t _port, unsigned int _proto, const char *_ip, int ipv6)
+int OS_Bindport(char *_port, unsigned int _proto, const char *_ip)
 {
-    int ossock;
-    struct sockaddr_in server;
+    int ossock = 0, s;
+    struct addrinfo hints, *result, *rp;
 
-#ifndef WIN32
-    struct sockaddr_in6 server6;
-#else
-    ipv6 = 0;
-#endif
 
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = AF_INET6;    /* Allow IPv4 or IPv6 */
+    hints.ai_protocol = _proto;
     if (_proto == IPPROTO_UDP) {
-        if ((ossock = socket(ipv6 == 1 ? PF_INET6 : PF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0) {
-            return OS_SOCKTERR;
-        }
+        hints.ai_socktype = SOCK_DGRAM;
     } else if (_proto == IPPROTO_TCP) {
-        int flag = 1;
-        if ((ossock = socket(ipv6 == 1 ? PF_INET6 : PF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
-            return (int)(OS_SOCKTERR);
-        }
-
-        if (setsockopt(ossock, SOL_SOCKET, SO_REUSEADDR,
-                       (char *)&flag,  sizeof(flag)) < 0) {
-            OS_CloseSocket(ossock);
-            return (OS_SOCKTERR);
-        }
+        hints.ai_socktype = SOCK_STREAM;
     } else {
-        return (OS_INVALID);
+        return(OS_INVALID);
     }
+    hints.ai_flags = AI_PASSIVE | AI_ADDRCONFIG | AI_V4MAPPED;
```

Comment:
If we're going to intentionally not supportmajor features on openbsd, why
not drop support for it? I wonder how other projecta manage to support both
ipv6 and ipv4 on openbsd.
On Mar 29, 2015 12:02 AM, "Antonio Querubin" <notifications@github.com>
wrote:

> In src/os_net/os_net.c
> <https://github.com/ossec/ossec-hids/pull/563#discussion_r27351181>:
>
> >      }
> > +    hints.ai_flags = AI_PASSIVE | AI_ADDRCONFIG | AI_V4MAPPED;
>
> On Tue, 24 Mar 2015, ddpbsd wrote: > } > + hints.ai_flags = AI_PASSIVE |
> AI_ADDRCONFIG | AI_V4MAPPED; The only place I see anything about
> AI_V4MAPPED in os_net.h is inside of a win32 block.
> Went down some blind alleys troubleshooting this as I'd forgotten how
> single-socket operations actually work on dual-stack hosts. It's also
> reminded me of a long-known but obscure problem with OpenBSD's support (or
> rather lack) of the socket API (RFC 3493). To make a long story short,
> OpenBSD doesn't allow binding to a single socket for both IPv4 and IPv6.
> Unlike other *BSDs, it isn't even allowed as a boot-time nor runtime
> configurable option without recompiling a whole new kernel. And even that's
> not so simple because the system headers are intentionally non-compliant
> with the RFC/API and missing key flags (eg. AI_V4MAPPED). The workaround I
> finally implemented allows OSSEC as a server to run on an OpenBSD system
> listening on IPv4. As an agent, it should work fine with any address mix on
> single-stack or dual-stack systems. Antonio Querubin e-mail:
> tony@lavanauts.org xmpp: antonioquerubin@gmail.com
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/ossec/ossec-hids/pull/563/files#r27351181>.
>